### PR TITLE
Fix popup detection

### PR DESCRIPTION
**Pull request by muodov**
This PR contains 3 things (which depend on each other):

1. Fix a bug with "visible" element detection, which was erroneously returning true in most cases, if the element had no "display" in its style attribute. This caused false positives on popup detection, which resulted in subsequent failures
2. prehideSelectors now use "opacity" instead of "display" to avoid subsequent conflicts with "visible" rules
3. prehideSelectors are implemented in playwright tests (supersedes #7)

### CODE REVIEW

File: lib/cmps/base.ts [hunk 1]
```diff
@@ -140,7 +140,9 @@ export class AutoConsent extends AutoConsentBase {
 
   async _runRulesSequentially(tab: TabActor, rules: AutoConsentRuleStep[]): Promise<boolean> {
     for (const rule of rules) {
+      console.log('Running rule...', rule, tab.id);
```
Comment (sammacbeth) [inline lib/cmps/base.ts:143]: You've added quite a bit of logging with `console.log` throughout this PR. I'd prefer if we trimmed this down a bit, and where we do have logging put it behind a flag, so it is off by default.

File: lib/node.ts [hunk 2]
```diff
@@ -11,11 +12,13 @@ export {
   TabConsent,
 }
 
-export function attachToPage(page: any, url: string, rules: AutoCMP[], retries = 1) {
-  const frames: { [id: number]: any } = {};
+export async function attachToPage(page: any, url: string, rules: AutoCMP[], retries = 1) {
```
Comment (sammacbeth) [inline lib/node.ts:15]: Making this async is a breaking API change which will affect downstream users (i.e. [tracker-radar-collector](https://github.com/duckduckgo/tracker-radar-collector/blob/main/collectors/CMPCollector.js)).

File: lib/node.ts [hunk 3]
```diff
@@ -34,5 +37,6 @@ export function attachToPage(page: any, url: string, rules: AutoCMP[], retries =
   }
   page.on('framenavigated', onFrame);
   page.frames().forEach(onFrame);
-  return new TabConsent(tab, detectDialog(tab, retries, rules))
+  await prehideElements(tab, rules);
```
Comment (sammacbeth) [inline lib/node.ts:40]: I would make this optional - i.e. the caller specifies if prehiding should be done or not (as it's an invasive change).

File: lib/puppet/tab.ts [hunk 4]
```diff
@@ -23,6 +25,18 @@ export default class Tab implements TabActor {
...
+  async _initPlaywrightUtils() {
+    const script = `(() => {
+      ${getStyleElementUtil.toString()}
+      ${hideElementsUtil.toString()}
+      window.autoconsentTestUtils = {
+        getStyleElementUtil,
+        hideElementsUtil,
+      };
+    })();`;
+    return this.frames[0].evaluate(script);
```
Comment (sammacbeth) [inline lib/puppet/tab.ts:37]: Can we not inject these functions on demand when we do the actual hiding? This prevents us from having to do injection when the page loads, and also means we don't have to attach things to global scope.

Note this also might break if the page navigates after we attached to it.

File: lib/cmps/base.ts (hunk 1 above)
Comment (muodov) [reply lib/cmps/base.ts:143]: Agreed. These messages are really helpful during debugging, but I noticed they were quite noisy in jenkins runs. For the user it shouldn't matter because these messages are printed in the background page

File: lib/node.ts (hunk 2 above)
Comment (muodov) [reply lib/node.ts:15]: oh. Thanks for the heads up, I didn't realize it was part of an API

File: lib/puppet/tab.ts [hunk 5]
```diff
@@ -101,9 +115,8 @@ export default class Tab implements TabActor {
     return false;
   }
 
-  async hideElements(selectors: string[], frameId = 0) {
-    // TODO implement this
-    return Promise.resolve(true)
+  async hideElements(selectors: string[], frameId = 0, method: HideMethod = 'display') {
+    return await this.frames[frameId].evaluate(`window.autoconsentTestUtils.hideElementsUtil(${JSON.stringify(selectors)}, '${method}')`);
```
[the remark this answers is not in the document]
Comment (muodov) [reply lib/puppet/tab.ts:119]: The idea is to have reusable functions that are called from the content script, and injected by the playwright harness. I think the other RPC commands can be refactored in the same way, to make test setup closer to the production version.

Since those functions may call into each other, they need to be either saved in globals, or injected in `evaluate()` call. I figured that  The edge case with navigation is a good point, so I guess we should inject them every time indeed, as long as we keep the background page model.

As for the function arguments, I tried passing them at first, but in this case the first argument has to be a function, otherwise Playwright always returns undefined as the result. If we inject multiple functions, their source has to be rendered in a single string, I don't think there's a way around that.

I'll poke around a bit more and see if I can get rid of `_initPlaywrightUtils()`

File: tests/runner.ts [hunk 6]
```diff
@@ -39,7 +39,7 @@ export function generateTest(url: string, expectedCmp: string, options: TestOpti
         }
         await page.goto(url, { waitUntil: 'commit' });
 
-        const tab = autoconsent.attachToPage(page, url, rules, 20);
+        const tab = await autoconsent.attachToPage(page, url, rules, 20, true);
```
Comment (sammacbeth) [inline tests/runner.ts:42]:
```suggestion
        const tab = autoconsent.attachToPage(page, url, rules, 20, true);
```
Await is no longer necessary.

File: rules/autoconsent/etsy.json [hunk 7]
```diff
@@ -0,0 +1,13 @@
+{
+    "name": "etsy",
```
Comment (sammacbeth) [inline rules/autoconsent/etsy.json:2]: Can you add a test for this rule?

**Review by sammacbeth — APPROVED** [review]
Looks great - just a couple of minor comments to address, then I think we're good to merge.
